src/hotspot/share/code/aotCodeCache.cpp

```diff
@@ -0,0 +1,1283 @@
+      current += size;
+      uint n = write_bytes(&(entries_address[i]), sizeof(AOTCodeEntry));
+      if (n != sizeof(AOTCodeEntry)) {
+        FREE_C_HEAP_ARRAY(uint, search);
+        return false;
+      }
+      search[entries_count*2 + 0] = entries_address[i].id();
+      search[entries_count*2 + 1] = entries_count;
+      entries_count++;
+      AOTCodeEntry::Kind kind = entries_address[i].kind();
+      if (kind == AOTCodeEntry::Adapter) {
+        adapters_count++;
+      } else if (kind == AOTCodeEntry::Blob) {
+        total_blobs_count++;
+      }
+    }
+    if (entries_count == 0) {
+      log_info(aot, codecache, exit)("AOT Code Cache was not created: no entires");
+      FREE_C_HEAP_ARRAY(uint, search);
+      return true; // Nothing to write
+    }
+    assert(entries_count <= store_count, "%d > %d", entries_count, store_count);
+    // Write strings
+    if (strings_count > 0) {
+      copy_bytes((_store_buffer + strings_offset), (address)current, strings_size);
+      strings_offset = (current - start); // New offset
+      current += strings_size;
+    }
+
+    uint new_entries_offset = (current - start); // New offset
+    // Sort and store search table
+    qsort(search, entries_count, 2*sizeof(uint), uint_cmp);
+    search_size = 2 * entries_count * sizeof(uint);
+    copy_bytes((const char*)search, (address)current, search_size);
+    FREE_C_HEAP_ARRAY(uint, search);
+    current += search_size;
+
+    // Write entries
+    entries_size = entries_count * sizeof(AOTCodeEntry); // New size
+    copy_bytes((_store_buffer + entries_offset), (address)current, entries_size);
+    current += entries_size;
+    log_info(aot, codecache, exit)("Wrote %d AOTCodeEntry entries (%d max size) to AOT Code Cache", entries_count, max_size);
+    log_info(aot, codecache, exit)("  Adapters:  total=%d", adapters_count);
+    log_info(aot, codecache, exit)("  All Blobs: total=%d", total_blobs_count);
+
+    uint size = (current - start);
+    assert(size <= total_size, "%d > %d", size , total_size);
+
+    // Finalize header
+    AOTCodeCache::Header* header = (AOTCodeCache::Header*)start;
+    header->init(size, (uint)strings_count, strings_offset,
+                 entries_count, new_entries_offset);
+    log_info(aot, codecache, exit)("Wrote %d bytes to AOT Code Cache", size);
+  }
+  return true;
+}
+
+//------------------Store/Load AOT code ----------------------
+
+bool AOTCodeCache::store_code_blob(CodeBlob& blob, AOTCodeEntry::Kind entry_kind, uint id, const char* name, int entry_offset_count, int* entry_offsets) {
+  AOTCodeCache* cache = open_for_dump();
+  if (cache == nullptr) {
+    return false;
+  }
+  assert(AOTCodeEntry::is_valid_entry_kind(entry_kind), "invalid entry_kind %d", entry_kind);
+
+  if ((entry_kind == AOTCodeEntry::Adapter) && !cache->adapter_caching()) {
+    return false;
+  }
+  log_info(aot, codecache, stubs)("Writing blob '%s' to AOT Code Cache", name);
```

Comment:
I'd revisit logging code in AOTCodeCache and downgrade info->debug and debug->trace where appropriate. It feels too low-level most of the time.

Reply:
done